integration/service/connectivitymap/builder.go

```diff
@@ -0,0 +1,183 @@
+
+	"github.com/Telenav/osrm-backend/integration/service/spatialindexer"
+	"github.com/golang/glog"
+)
+
+type connectivityMapBuilder struct {
+	iterator      spatialindexer.PointsIterator
+	finder        spatialindexer.Finder
+	ranker        spatialindexer.Ranker
+	distanceLimit float64
+	id2NearbyIDs  ID2NearByIDsMap
+
+	numOfWorker         int
+	dispatchWaitGroup   *sync.WaitGroup
+	workerWaitGroup     *sync.WaitGroup
+	aggregatorWaitGroup *sync.WaitGroup
+	tasks4WorkerC       []chan spatialindexer.PointInfo
+	aggregatorC         chan placeIDWithNearByPlaceIDs
+}
+
+func newConnectivityMapBuilder(iterator spatialindexer.PointsIterator, finder spatialindexer.Finder,
+	ranker spatialindexer.Ranker, distanceLimit float64, numOfWorker int) *connectivityMapBuilder {
+	builder := &connectivityMapBuilder{
+		iterator:      iterator,
+		finder:        finder,
+		ranker:        ranker,
+		distanceLimit: distanceLimit,
+		id2NearbyIDs:  make(ID2NearByIDsMap),
+
+		numOfWorker:         numOfWorker,
+		dispatchWaitGroup:   &sync.WaitGroup{},
+		workerWaitGroup:     &sync.WaitGroup{},
+		aggregatorWaitGroup: &sync.WaitGroup{},
+		tasks4WorkerC:       make([]chan spatialindexer.PointInfo, numOfWorker),
+		aggregatorC:         make(chan placeIDWithNearByPlaceIDs, 10000),
+	}
+
+	if numOfWorker < 1 {
+		glog.Fatal("numOfWorker should never be smaller than 1, recommend using NumCPU()\n")
+	}
+
+	for i := range builder.tasks4WorkerC {
+		builder.tasks4WorkerC[i] = make(chan spatialindexer.PointInfo, 50)
+	}
+
+	return builder
+}
+
+/*
+                                    ->   tasksChannel   ---> worker
+                                  /                                   \
+                                 /                                     \
+Input Iterator ==> dispatcher    --->    tasksChannel   ---> worker      ---> aggregatorChannel -> iterate and put result to map
+                                 \                                     /
+                                  \                                   /
+                                    ->   tasksChannel   ---> worker
+*/
+
+func (builder *connectivityMapBuilder) build() ID2NearByIDsMap {
+	builder.dispatch()
+	builder.process()
+	builder.aggregate()
+	builder.wait()
+
+	return builder.id2NearbyIDs
+}
+
+func (builder *connectivityMapBuilder) dispatch() {
+	builder.dispatchWaitGroup.Add(1)
+
+	go func() {
+		counter := 0
+		for p := range builder.iterator.IteratePoints() {
+			builder.tasks4WorkerC[counter%builder.numOfWorker] <- p
+			counter += 1
+		}
+
+		for i := 0; i < builder.numOfWorker; i++ {
+			close(builder.tasks4WorkerC[i])
+		}
+
+		glog.Infof("builder::dispatch is finished.  Total input is %d.\n", counter)
+		builder.dispatchWaitGroup.Done()
+	}()
+}
+
+func (builder *connectivityMapBuilder) process() {
+	for i := 0; i < builder.numOfWorker; i++ {
+		builder.workerWaitGroup.Add(1)
+		go builder.work(builder.workerWaitGroup, i, builder.tasks4WorkerC[i], builder.aggregatorC)
```

Comment:
Fully agree on the idea.  I am also thinking about this trade-off during this part of code, which is, 
- let `dispatcher()` dispatch incoming tasks and each `worker` just need to handle his own,  the downside is there will be new object memory be allocated for each of them, we could double the input's memory
- another way is, put the dispatch related logic inside `worker`, each worker just handle `currentIndex % totalNumTasks == currentTaskID`, it just iterate data for more times but I think this is quick

Today, before I start to make the change, I thought another trade-off I have made, which is the input of `iterator` part. I use the `go-channel` to implement this iterator, which means source data will be put into channel to let others use it, which makes here must use `dispatcher()`.  I think the benefit is, at `producer` side to prepare this `iterator`, it creates the channel and immediately return it, at `consumer` side it will holds if `iterator` is not finished.  If some one calls when `iterator` is not ready, then he just wait there.

If I choose to implement `iterator` by additional `iterator` class which implements `next()` and `hasnext()`, then for upper part's logic, it needs **additional attention** to duplicate iterator itself and send different instance to different `go-routine`.

I also have a think about `go's` iterator, when to use `channel` and when to use traditional way
- If the data you want to iterate is ready or you are sure that `iterator` will only be called when data has already be cooked, then traditional way of `iterator struct` + `next()` + `hasNext()` should work for most situations.  
- If the target object have some heavier task to prepare(file I/O, network, heavy calculation, etc), and users are impatient to hold the `iterator`, then channel might be good for this.
- In general, I think traditional way is more simpler and suitable for majority of cases.  
- For this part, I will keep current iterator for now but will consider whether I could simplify it.

Reply:
After sync with @wangyoucao577, I got his point and I think his suggestion is better: the logic of `dispatcher` is unnecessary, just pass `input channel` into each of `worker`, they could directly read from `channel`, let `go-routine`'s scheduler do the dispatch: the worker who get chance to run will fetch item from channel.

```
                               ->  worker (fetch task  ->  find  ->  rank)
                             /                                               \
                            /                                                 \
  Input Iterator(channel)    --->  worker (fetch task  ->  find  ->  rank)      ---> aggregatorChannel -> feed to map
                            \                                                 /
                             \                                               /
                               ->  worker (fetch task  ->  find  ->  rank)
```

Here is the change:
https://github.com/Telenav/osrm-backend/pull/259/commits/3a9474fe58433af61cb506c887fbd8a7758c4f8a